### ISSUE TITLE
Error in estimating statistics for ragged arrays

### ISSUE THREAD

**Issue by mikemeredith** [kenkellner/jagsUI#49]
A preliminary heads-up: With the current devel version 1.5.2.9001, `jags` throws an error at the "Calculating statistics..." stage when the model includes a ragged array. The error message is
`Error in quantile.default(x, y) : 
  missing values and NaN's not allowed if 'na.rm' is FALSE`

This happens in several models in the IPM book code (Schaub & Kery 2022), starting with section 10.6. There we have a 4 x 5 matrix `rho` for a correlation coefficient, where only the elements above the diagonal are defined, so a ragged array.

I think this is due to changes in the "Don't drop chains with all NA" commit.

The immediate error can be fixed by inserting `na.rm=TRUE` in the call to `quantile `in line 28 of process.output.R. We also need to fix `gf `in line 32 by inserting `if(is.na(mean(x)))   return(NA)`.

Then I get `Error in '.rowNamesDF<-'(x, value = value) : invalid 'row.names' length` which is coming from line 49 in summary.matrix.R. I've not managed to sort that out, but there seems to be confusion because `process.output` is removing stuff in `codaOnly `and `summary.matrix` is trying to do that too - as well as removing all-NA columns - so we end up with a mismatch between `p[!expand%in%codaOnly]` and the rows of `y`.

I think the strategy is to have `process.output` do everything necessary to remove `codaOnly` parameters, then have `summary.matrix` just build the matrix and then remove the all-NA rows.

I'll work on it some more and hopefully submit a pull request in a day or 2.

**Comment by kenkellner** [comment]
Clearly I didn't think this through. Maybe it would be better to revert the change for now until we come up with a good overall solution? I think avoiding dropping NA columns is only needed in very unusual modeling situations and only when running in parallel. Maybe we could just provide the option as an argument for those specific situations.

**Comment by mikemeredith** [comment]
As I understand it, the problem with using `na.rm=TRUE` (the default) in the call to `coda.samples` (runmodel.R #95) occurs when different chains have different nodes which are all-NA. Running in parallel, the `mcmc` objects in runparallel.R #62 have different nodes and `as.mcmc.list` fails.

One strategy might be to have `na.rm=FALSE` in the call to `coda.samples` when running in parallel, then remove all-NA nodes after creating the `mcmc.list` object.

This could be done after creating `mat` in process.output.R #19, and then using `params <- colnames(mat)` instead of `colnames(x[[1]])` (line 13). (Actually, would suggest using `node.names` instead of `params` for this object.)

In summary.matrix.R #28, we go back to `samples` and redo the node names thing all over again - and sometimes get it wrong! Suggest including `params` (or `node.names`) in the output from `process.output` and using that in `summary.matrix`. You would still need to remove the `codaOnly` parameters from the row names of `y`, but not have to worry about all-NA nodes.

**Comment by kenkellner** [comment]
6da27bb should fix this, I think. Columns of NAs are kept only when running in parallel. The chains are then joined into an `mcmc.list`, and any parameter entirely consisting of NAs is dropped from the list. This happens before `process.output`. This should avoid the original problem of different nodes being NA, while minimizing changes to other functions.

I tested it on a toy model that I forced to have all NA parameters and it works. I  don't know of a model structure that will reliably create the original condition (some chains with NA some without). I'll see  if I can get the model from the original bug report to run.

**Comment by mikemeredith** [comment]
Thanks Ken, that works with my ragged arrays now and with the cases where I use `codaOnly`.
